### DISCUSSION TITLE
Parser backtracking / greedyness

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1754] [Q&A]
Given a grammar

```
QUALIFIED_PATH returns string: ID ("." ID)*;
FieldTypeLike:
    'field''<' node=QUALIFIED_PATH '.' field=ID '>';
```

QUALIFIED_PATH will always eat up the complete FQN on e.g.

`field<a.b.Myclass.myfield>`

is there a way to tell chevrotain to backtrack?

**Comment by msujew** [comment]
Hey @cdietrich,

in theory yes, a `GATE` on the `MANY` would allow to write a syntactic predicate in order to prevent greedy parsing. However, the current API of Langium doesn't expose this feature of Chevrotain. It's pretty locked down.

In the future I hope to refactor a lot of the parsing mechanism, see also https://github.com/eclipse-langium/langium/issues/1742. Exposing more of the existing Chevrotain API is also part of that.

**Reply by cdietrich** [reply]
ok so that would then required two QUALIFIED_PATH rules. one with a gate that rejects `"." ID >` and one for other places?

**Reply by msujew** [reply]
Well, it depends on your use case. But yes, in that you either need two rules, or a parameterized rule.

**Reply by cdietrich** [reply]
Is there a way to get a chrvrotain rule for a given langium rule ?

**Reply by msujew** [reply]
As said, it's pretty locked down. See [here](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/parser/parser-builder-base.ts).
